File: mps-cli-rs/src/model/snoderef.rs
use crate::model::snode::SNode;
use crate::model::srepository::SRepository;
use std::rc::Rc;

pub struct SNodeRef {
    pub model_id: String,
    pub node_id: String,
    pub resolve_info: String,
    pub referenced_node : Option<Rc<SNode>>,
}

impl SNodeRef {
    pub fn new(model_id: String, node_id: String, resolve_info: String) -> Self {
        SNodeRef {
            model_id,
            node_id,
            resolve_info,
            referenced_node : None,
        }
    }

    pub fn resolve(&self, repository: &SRepository) -> Option<Rc<SNode>> {
        if self.referenced_node.is_some() {
            return self.referenced_node.clone();
        }

        return match repository.get_model_by_uuid(self.model_id.as_str()) {
            None => None,
            Some(model) => 
                if let Some(n) = model.borrow().get_node_by_id(self.node_id.as_str()) {
                    return Some(n);
                } else {
                    return None;
                }
        };
    }
}

File: mps-cli-py/src/mpscli/model/SNodeRef.py
class SNodeRef:
    def __init__(self, model_uuid, node_uuid, resolve_info=None):
        self.model_uuid = model_uuid
        self.node_uuid = node_uuid
        self.resolve_info = resolve_info

    def resolve(self, repo):
        model = repo.get_model_by_uuid(self.model_uuid)
        if model is None:
            return None
        return model.get_node_by_uuid(self.node_uuid)
